src/approx_inverse_retraction.jl

```diff
@@ -0,0 +1,103 @@
+"""
+    ApproximateInverseRetraction <: AbstractInverseRetractionMethod
+
+An abstract type for representing approximate inverse retraction methods.
+"""
+abstract type ApproximateInverseRetraction <: AbstractInverseRetractionMethod end
+
+"""
+    NLsolveInverseRetraction{T<:AbstractRetractionMethod} <: ApproximateInverseRetraction
+
+An inverse retraction method for approximating the inverse of a retraction using `NLsolve`.
+
+# Constructor
+
+    NLsolveInverseRetraction(
+        method::AbstractRetractionMethod[, X0];
+        project_tangent=false,
+        project_point=false,
+        nlsolve_kwargs...,
+    )
+
+Constructs an approximate inverse retraction for the retraction `method` with initial guess
+`X0`, defaulting to the zero vector. If `project_tangent` is `true`, then the tangent
+vector is projected before the retraction using `project`. If `project_point` is `true`,
+then the resulting point is projected after the retraction. `nlsolve_kwargs` are keyword
+arguments passed to `NLsolve.nlsolve`.
+"""
+struct NLsolveInverseRetraction{TR<:AbstractRetractionMethod,TV} <:
+       AbstractInverseRetractionMethod
+    retraction::TR
+    X0::TV
+    project_tangent::Bool
+    project_point::Bool
+    nlsolve_kwargs
```

Comment:
Why is this field abstractly typed?

Reply:
Because they're used as kwargs, their type can be anything that expands correctly when splatted, such as an `AbstractDict`, iterator of `Pair`s, or `NamedTuple`, but potentially other types. We don't include the type in the signature, because, e.g. if the user provides a `NamedTuple`, then `inverse_retract` would specialize on the kwargs provided, even though `nlsolve` itself would not specialize on these kwargs. I _think_ this is the right thing to do here.

Reply:
As far as I know `nlsolve` itself would actually specialize on that (unless you pass functions or types but that should not be the case). I just tried a simple benchmark:
```julia
julia> struct A{T}
           kwargs::T
       end

julia> using NLsolve

julia> f(t) = sin(t[1])
f (generic function with 2 methods)

julia> hh(a::A) = nlsolve(f, [2.0]; a.kwargs...)
hh (generic function with 1 method)

julia> a2 = A((; xtol=1e-12))
A{NamedTuple{(:xtol,),Tuple{Float64}}}((xtol = 1.0e-12,))

julia> using BenchmarkTools

julia> @benchmark hh($a2)
```
and it's 5.0 us for concretely typed `kwargs` vs 5.2 us for abstractly typed `kwargs` so it's really hard to tell what kind of performance impact it could have, especially when (in the abstractly typed case) the result of optimization is also abstractly typed.

Reply:
I changed the kwargs to be concretely typed.